File: app/src/main/java/cn/intret/app/picgo/ui/main/move/MoveFilePresenter.java
package cn.intret.app.picgo.ui.main.move;

import android.arch.lifecycle.LifecycleOwner;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.List;

import cn.intret.app.picgo.model.image.ImageModule;
import cn.intret.app.picgo.ui.base.BaseLifecyclePresenter;
import cn.intret.app.picgo.utils.RxUtils;

/**
 * Created by intret on 2018/3/17.
 */

public class MoveFilePresenter<V extends MoveFileContract.View & LifecycleOwner>
        extends BaseLifecyclePresenter<V> implements MoveFileContract.Presenter {

    V mView;

    public MoveFilePresenter(@NotNull V view) {
        mView = view;
    }

    @Override
    public void loadFolderList() {
        ImageModule.getInstance()
                .loadFolderList(true)
                .compose(RxUtils.applySchedulers())
                .as(RxUtils.lifecycleDisposable(mView))
                .subscribe(folderModel -> mView.onLoadFolderModelSuccess(folderModel),
                        throwable -> {
                            mView.onLoadFolderModelFailed();
                        });
    }

    @Override
    public void detectFileExistence(List<File> sourceFiles) {
        ImageModule.getInstance()
                .detectFileExistence(sourceFiles)
                .compose(RxUtils.applySchedulers())
                .as(RxUtils.lifecycleDisposable(mView))
                .subscribe(detectFileExistenceResult -> {
                    mView.onDetectFileExistenceResult(detectFileExistenceResult);
                }, throwable -> {
                    mView.onDetectFileExistenceFailed(sourceFiles);
                });
    }


    @Override
    public void start() {

    }
}

File: app/src/main/java/cn/intret/app/picgo/ui/main/move/MoveFilePresenter.kt
package cn.intret.app.picgo.ui.main.move

import android.arch.lifecycle.LifecycleOwner
import cn.intret.app.picgo.model.image.DetectFileExistenceResult
import cn.intret.app.picgo.model.image.FolderModel
import cn.intret.app.picgo.model.image.ImageModule
import cn.intret.app.picgo.ui.base.BaseLifecyclePresenter
import cn.intret.app.picgo.utils.RxUtils
import java.io.File

/**
 * Presenter role of MoveFile dialog's MPV design pattern
 */

class MoveFilePresenter<V>(var mView: V)
    : BaseLifecyclePresenter<V>(), MoveFileContracts.Presenter
        where V : MoveFileContracts.View, V : LifecycleOwner {

    override fun loadFolderList() {

        var lifecycleDisposable =
        ImageModule
                .loadFolderModel(true)

                .compose(RxUtils.applySchedulers())
                .`as`(RxUtils.lifecycleDisposable(mView))

                .subscribe(
                        { folderModel: FolderModel -> mView.onLoadFolderModelSuccess(folderModel) },
                        { throwable: Throwable -> mView.onLoadFolderModelFailed() })
    }

    override fun detectFileExistence(sourceFiles: List<File>?) {
        sourceFiles?.let {
            ImageModule
                    .detectFileExistence(it)
                    .compose<DetectFileExistenceResult>(RxUtils.applySchedulers<DetectFileExistenceResult>())
                    .`as`(RxUtils.lifecycleDisposable<DetectFileExistenceResult>(mView))
                    .subscribe(
                            { detectFileExistenceResult: DetectFileExistenceResult ->
                                mView.onDetectFileExistenceResult(detectFileExistenceResult)
                            },
                            { throwable: Throwable ->
                                mView.onDetectFileExistenceFailed(sourceFiles)
                            })

        }
    }


    override fun start() {

    }
}
